File: Link_Prediction/PAC.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    vector<vector<int>> adj = {{2, 3, 6},{3, 4, 6},{0, 3},{0, 1, 2, 4, 5},{1, 3, 5, 7},{3, 4, 7},{0, 1},{4, 5}};
    int n = adj.size();
    vector<pair<int,int>> node_pairs;
    vector<int> degree(n);

    for(int i=0; i<n; i++){
        for(int j=i+1; j<n; j++){
            if(find(adj[i].begin(), adj[i].end(), j) == adj[i].end()){
                node_pairs.push_back({i,j});
            }
        }
    }

    for(int i=0; i<n; i++){
        degree[i] = adj[i].size();
    }

    int res = 0;
    vector<pair<int,int>> ans;
    for(auto& it: node_pairs){
        int pac = degree[it.first] * degree[it.second];
        if(pac > res){
            res = pac;
            ans = {{it.first, it.second}};
        }
        else if(pac == res){
            ans.push_back({it.first, it.second});
        }
        cout << "(" <<it.first << "," << it.second << "): " << pac << endl;
    }

    cout << "Maximum Product of Degrees: " << res << endl;
    cout << "Predicted Link: " << endl;
    for(auto& it: ans){
        cout << it.first << "-" << it.second << endl;
    }
}

File: Link_Prediction/PAC.py
adj = [[2, 3, 6], [3, 4, 6], [0, 3], [0, 1, 2, 4, 5], [1, 3, 5, 7], [3, 4, 7], [0, 1], [4, 5]]
n = len(adj)
node_pairs = []
degree = [0] * n

#finding all pairs of nodes that are not connected
for i in range(n):
    for j in range(i + 1, n):
        if j not in adj[i]:
            node_pairs.append((i, j))

for i in range(n):
    degree[i] = len(adj[i])

res = 0
ans = []

# calculating PAC for each pair of nodes that are not connected
for u, v in node_pairs:
    pac = degree[u] * degree[v]
    if pac > res:
        res = pac
        ans = [(u, v)]
    elif pac == res:
        ans.append((u, v))
    print(f"({u},{v}): {pac}")

print("Maximum Product of Degrees:", res)
print("Predicted Link:")
for u, v in ans:
    print(f"{u}-{v}")